### PR TITLE
bump controller-tools to v0.2.0

### PR DESCRIPTION
**Pull request by estroz**
**Description of the change:** bump controller-tools to v0.2.0 and update CRD generator API/unit tests.

**Motivation for the change:** controller-tools k8s deps versions should match those of controller-runtime when possible. #1143 requires some changes in this release too.

Note: #1607 should be closed once the refactor base branch is merged with master.

### CODE REVIEW

File: internal/pkg/scaffold/crd.go [hunk 1]
```diff
@@ -58,58 +63,51 @@ func (s *CRD) getFS() afero.Fs {
...
+}
+
+type crdOutputRule struct {
+	fs afero.Fs
+}
+
+var _ genall.OutputRule = crdOutputRule{}
+
+// Open is meant to be used to generate a CRD manifest in memory at path.
+func (o crdOutputRule) Open(_ *loader.Package, path string) (io.WriteCloser, error) {
```
Comment (LiliC) [inline internal/pkg/scaffold/crd.go]: Where is this used?

File: internal/pkg/scaffold/crd_test.go [hunk 2]
```diff
@@ -65,47 +65,48 @@ spec:
...
-              items:
-                type: string
-              type: array
-          required:
-          - nodes
-          type: object
-  version: v1alpha1
   versions:
   - name: v1alpha1
+    schema:
```
Comment (LiliC) [inline internal/pkg/scaffold/crd_test.go]: So if I understand correctly the CRD API changed here, that's why the changes?

File: internal/pkg/scaffold/crd.go (hunk 1 above)
Comment (estroz) [reply internal/pkg/scaffold/crd.go]: The CRD generator will call `o.Open()` when writing the CRD manifest because it is an [`OutputRule`](https://github.com/operator-framework/operator-sdk/pull/1642/files#diff-88638833fc541d7d24b44b34835b0fb4R168).

File: internal/pkg/scaffold/crd_test.go (hunk 2 above)
Comment (estroz) [reply internal/pkg/scaffold/crd_test.go]: Yes now that `version` is deprecated, the CRD generator will not create a `version` field or global validation by default.

**Comment by estroz** [comment]
Two changes that need discussion:

1. CRD manifests can contain multiple specified versions, so file naming should change from `<group>_<version>_<kind>.yaml` to something else. For example, `controller-tools` uses `<group>_<resource>.yaml`.
1. The CRD generator permits legacy support for `spec.version` but not simultaneously with `spec.versions`. The `checkCRDVersions()` function permits both forms as long as `spec.version` is the first element in `spec.versions`, which is correct [behavior](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#overview). How should we handle this?

**Comment by hasbro17** [comment]
@estroz Can we rebase or cherry pick #1652 into the `refactor/controller-runtime-v0.2.0` branch?
And update the tag from `// +genclient:nonNamespaced` to `// +kubebuilder:resource:path=<lowercase-kind-plural>,scope=Namespaced`

Or do you want to do that as a separate PR?

**Comment by LiliC** [comment]
@estroz For .1 I think we should follow the convention that kubebuilder has, so it will make it easier to migrate, do they have anything established already?

For 2. I think the log about the deprecation and error out if spec is not met, sounds good to me.

**Comment by estroz** [comment]
CRD file naming changes were quite extensive and touched both Helm and Ansible docs/tests. PTAL.

/cc @LiliC @hasbro17 @AlexNPavel @joelanford @jmrodri @fabianvf

**Comment by estroz** [comment]
@LiliC kubebuilder would generate a CRD file name `cache.example.com_memcacheds.yaml`. IMO `cache_memcacheds_crd.yaml` looks much nicer and contains all the necessary uniqueness/information that the kubebuilder name gives. WDYT?

File: CHANGELOG.md [hunk 3]
```diff
@@ -26,6 +26,8 @@
 - Upgrade [`controller-runtime`](https://github.com/kubernetes-sigs/controller-runtime) version from `v0.1.12` to `v0.2.0-beta.3`
 - Upgrade Kubernetes version from `kubernetes-1.13.4` to `kubernetes-1.14.1`
 - Upgrade `github.com/operator-framework/operator-lifecycle-manager` version from `b8a4faf68e36feb6d99a6aec623b405e587b17b1` to `0.10.1`
+- CRD file names were previously of the form `<group>_<version>_<kind>_crd.yaml`. Now that CRD manifest `spec.version` is deprecated in favor of `spec.versions`, i.e. multiple versions can be specified in one CRD, CRD file names have the form `<group>_<resource>_crd.yaml`. `<resource>` is the plural lower-case CRD Kind found at `spec.names.plural`.
```
Comment (hasbro17) [inline CHANGELOG.md]: Changing from `<group>_<version>_<kind>_crd.yaml` to `<group>_<resource>_crd.yaml` isn't a breaking change?

Will any existing CLI commands like `csv-gen` or `generate openapi` now break if users have manifest in the old format `<group>_<version>_<kind>_crd.yaml`?

**Comment by LiliC** [comment]
> @LiliC kubebuilder would generate a CRD file name cache.example.com_memcacheds.yaml. IMO cache_memcacheds_crd.yaml looks much nicer and contains all the necessary uniqueness/information that the kubebuilder name gives. WDYT?

Since we are moving to kubebuilder at some point, I would suggest we go the kubebuilder way as we probably will have to change it then anyways.

File: CHANGELOG.md (hunk 3 above)
Comment (estroz) [reply CHANGELOG.md]: It is indeed a breaking change for our testing infrastructure. `generate` commands will not break (`openapi` will write new CRD files), while `gen-csv` will only break if the CSV config contains specific file names and not just `deploy/crds`.

Regardless this should be moved to the "Breaking Changes". I will also move the dependency upgrades.

Comment (hasbro17) [reply CHANGELOG.md]: @estroz Is this already updated to be in the breaking changes section?

Also is it possible to rebase this `refactor/controller-runtime-v0.2.0` branch from the master since this CHANGELOG is now for v0.9.0.

Or do you want to hold off for this PR and do it later?

Comment (estroz) [reply CHANGELOG.md]: I haven't updated it yet, was waiting for more responses to comments. I'll update it now and rebase.

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,35 @@
...
+- Upgrade `github.com/operator-framework/operator-lifecycle-manager` version from `b8a4faf68e36feb6d99a6aec623b405e587b17b1` to `0.10.1`
+- Upgrade [`controller-runtime`](https://github.com/kubernetes-sigs/controller-runtime) version from `v0.1.12` to `v0.2.0-beta.3`
+  - The package `sigs.k8s.io/controller-runtime/pkg/runtime/scheme` is deprecated, and contains no code. Replace this import with `sigs.k8s.io/controller-runtime/pkg/scheme` where relevant.
+  - The package `sigs.k8s.io/controller-runtime/pkg/runtime/log` is deprecated. Replace this import with `sigs.k8s.io/controller-runtime/pkg/log` where relevant.
+  - The package `sigs.k8s.io/controller-runtime/pkg/runtime/signals` is deprecated. Replace this import with `sigs.k8s.io/controller-runtime/pkg/manager/signals` where relevant.
+  - [`sigs.k8s.io/controller-runtime/pkg/client.Client`](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/interfaces.go#L61)'s `List()` method has been updated: `List(ctx context.Context, opts *client.ListOptions, list runtime.Object) error` is now `List(ctx context.Context, list runtime.Object, opts ...client.ListOptionFunc) error`. To migrate:
```
Comment (hasbro17) [inline CHANGELOG.md]: Maybe link to the godocs for List()?
https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/client#Reader

Or use a commit SHA  instead of master in the link since the code could move around.

Comment (estroz) [reply CHANGELOG.md]: This change will go in a separate PR to the recator branch.

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,35 @@
...
+
+      ...
+
+      listOpts := &client.ListOptions{}
+      // Old
+      err = r.client.List(context.TODO(), listOps, podList)
+      // New
+      err = r.client.List(context.TODO(), podList, client.UseListOptions(listOps))
+      ```
+- CRD file names were previously of the form `<group>_<version>_<kind>_crd.yaml`. Now that CRD manifest `spec.version` is deprecated in favor of `spec.versions`, i.e. multiple versions can be specified in one CRD, CRD file names have the form `<full group>_<resource>.yaml`. `<resource>` is the plural lower-case CRD Kind found at `spec.names.plural`.
```
Comment (hasbro17) [inline CHANGELOG.md]: Per offline discussion we'll keep the `_crd` suffix to keep the naming congruent with the CR manifest: `<full group>_<resource>_crd.yaml`

File: test/ansible-inventory/molecule/test-local/playbook.yml [hunk 6]
```diff
@@ -24,7 +24,7 @@
     deploy_dir: "{{ lookup('env', 'MOLECULE_PROJECT_DIRECTORY') }}/deploy"
     pull_policy: Never
     REPLACE_IMAGE: inventory.apps.fabianism.us/inventory:testing
-    custom_resource: "{{ lookup('file', '/'.join([deploy_dir, 'crds/inventory_v1alpha1_inventory_cr.yaml'])) | from_yaml }}"
+    custom_resource: "{{ lookup('file', '/'.join([deploy_dir, 'crds/inventory.apps.fabianism.us_v1alpha1_inventory_cr.yaml'])) | from_yaml }}"
```
Comment (joelanford) [inline test/ansible-inventory/molecule/test-local/playbook.yml]: Nit and unrelated: should we change the domain to `example.com`?

Comment (estroz) [reply test/ansible-inventory/molecule/test-local/playbook.yml]: @fabianvf ok with this? we can change it in a follow-up PR

File: internal/pkg/scaffold/crd.go [hunk 7]
```diff
@@ -141,10 +146,64 @@ func (s *CRD) CustomRender() ([]byte, error) {
...
+
+	gctx := &genall.GenerationContext{
+		Collector: &markers.Collector{
+			Registry: &markers.Registry{},
+		},
+		Checker:    &loader.TypeChecker{},
+		InputRule:  genall.InputFromFileSystem,
+		OutputRule: crdOutputRule{fs: crdFS},
+	}
+	apiDir := filepath.Join(s.AbsProjectPath, ApisDir, r.GoImportGroup, strings.ToLower(r.Version))
```
[the remark this answers is not in the document]
Comment (estroz) [reply internal/pkg/scaffold/crd.go]: Good catch. In fact generating a CRD with these changes will overwrite the version in `spec.versions`. I'll play around with changing `apiDir`.

File: internal/pkg/scaffold/crd.go [hunk 8]
```diff
@@ -141,10 +145,70 @@ func (s *CRD) CustomRender() ([]byte, error) {
...
+	}
+
+	g := crdgen.Generator{}
+	if err := g.RegisterMarkers(gctx.Collector.Registry); err != nil {
+		return errors.Wrap(err, "error registering markers for CRD API versions")
+	}
+	if err := g.Generate(gctx); err != nil {
+		return errors.Wrap(err, "error generating a CRD for API versions")
+	}
+	return nil
```
Comment (joelanford) [inline internal/pkg/scaffold/crd.go]: Would it simplify this function to use [`Generators.ForRoots()`](https://github.com/kubernetes-sigs/controller-tools/blob/a653df4cae21d140ed307073ccbfb1d7e245d3a6/pkg/genall/genall.go#L130) and then run [this loop from `Runtime.Run()`](https://github.com/kubernetes-sigs/controller-tools/blob/a653df4cae21d140ed307073ccbfb1d7e245d3a6/pkg/genall/genall.go#L164-L170)

As a follow-up, maybe we should extract that loop into a separate `RunUntilError()` function upstream so we don't have to duplicate it?

File: internal/pkg/scaffold/go_mod.go [hunk 9]
```diff
@@ -61,7 +60,7 @@ replace (
 replace (
 	github.com/coreos/prometheus-operator => github.com/coreos/prometheus-operator v0.31.1
 	sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.2.0-beta.3
-	sigs.k8s.io/controller-tools => sigs.k8s.io/controller-tools v0.1.11-0.20190411181648-9d55346c2bde
+	sigs.k8s.io/controller-tools => sigs.k8s.io/controller-tools v0.2.0-beta.3
```
Comment (joelanford) [inline internal/pkg/scaffold/go_mod.go]: Do Go operator projects use controller-tools? Or can this be removed?

Ditto for `Gopkg.toml`

File: internal/pkg/scaffold/crd.go [hunk 10]
```diff
@@ -119,12 +117,18 @@ func (s *CRD) CustomRender() ([]byte, error) {
 		}
 		// controller-tools does not set ListKind or Singular names.
 		setCRDNamesForResource(crd, s.Resource)
```
[the remark this answers is not in the document]
Comment (estroz) [reply internal/pkg/scaffold/crd.go]: Yes it could. I will make a PR.
